### DISCUSSION TITLE
How to set error message when using custom validation function?

### DISCUSSION THREAD

**Discussion by msageryd** [react-hook-form/react-hook-form#7565] [Q&A] [answered]
I have searched discussions and docs for some hints on how to provide an errro message in custom validation functions. As per the docs a validation function should only return true or false.

I'm guessing that I'm supposed to `setError`? The buildt-in validations (required, max, etc) can take an optional error message. I figured that there might be a similar concept for custom valitators.

**Comment by Moshyfawn** [comment, marked as answer]
You can set your error messages in the custom validate function by returning them. See [this CSB](https://codesandbox.io/s/custom-validation-messages-zrkiu)
```
{
  required: 'First name field is required',
  validate: value => {
    if (value.length < 3) {
      return 'First name should contain more than 3 characters'
    }
    if (hasNumber(value)) {
      return "First name should't contain numbers"
    }

    return true
  },
}
```

**Reply by msageryd** [reply]
Thanks, that simplifies a lot.
I might have missed it, but I could not find this anywhere in the docs.

**Reply by Moshyfawn** [reply]
Huh, I don't see any explicit information about returning validation error messages either.
@bluebill1049 do you want to include some additional information to `validate` method API docs?

**Reply by bluebill1049** [reply]
<img width="953" alt="Screen Shot 2022-01-18 at 9 04 20 am" src="https://user-images.githubusercontent.com/10513364/149841730-45f4364f-a54d-41ae-b73a-2c90b3e9841e.png">

hey @Moshyfawn maybe I should redesign this with a switch instead of a radio button, especially when the user scroll past the top. i will think about it.

**Reply by msageryd** [reply]
Oh, I missed that completely. Sorry for being a sloppy reader.

**Reply by bluebill1049** [reply]
not your fault... i will think about a redesign and see if i can apply it in the following weeks.

**Reply by Moshyfawn** [reply]
Hi, @bluebill1049! I actually don't feel like the 2 sections _(with validation & with validation and message)_ should be split, tbh. Like, I'd excpect an example with multiple inputs returning different things in `validate`; `true` / `false` based on a condition, a validation message string. Maybe throw in an HTML attribute with a validation message as well.
Just thoughts, not forcing my opinions ;)

**Reply by bluebill1049** [reply]
oh, I see @Moshyfawn would like to send a PR to include that? 🙏

**Reply by Moshyfawn** [reply]
Sure, I can try, yeah!

**Reply by bluebill1049** [reply]
> Sure, I can try, yeah!

Sweet! It's in the documentation repo. Let me know if u need anything or help.
